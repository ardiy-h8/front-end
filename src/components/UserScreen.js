import React, { Component } from 'react'
import { TextField, Button } from 'material-ui'

import Navigation from './Navigation'
import Header from './Header'

const THREEx = THREEx || {}

THREEx.ArPatternFile = {}

THREEx.ArPatternFile.encodeImageUrl = (imageURL, onComplete) => {
  var image = new Image()
  image.onload = function () {
    var patternFileString = THREEx.ArPatternFile.encodeImage(image)
    onComplete(patternFileString)
  }
  image.src = imageURL
  image.onload()
}

THREEx.ArPatternFile.encodeImage = function (image) {
  var canvas = document.createElement('canvas')
  var context = canvas.getContext('2d')
  canvas.width = 16
  canvas.height = 16

  var patternFileString = ''
  for (
    var orientation = 0;
    orientation > -2 * Math.PI;
    orientation -= Math.PI / 2
  ) {
    context.save()
    context.clearRect(0, 0, canvas.width, canvas.height)
    context.translate(canvas.width / 2, canvas.height / 2)
    context.rotate(orientation)
    context.drawImage(
      image,
      -canvas.width / 2,
      -canvas.height / 2,
      canvas.width,
      canvas.height
    )
    context.restore()

    var imageData = context.getImageData(0, 0, canvas.width, canvas.height)
    if (orientation !== 0) patternFileString += '\n'
    for (var channelOffset = 2; channelOffset >= 0; channelOffset--) {
      for (var y = 0; y < imageData.height; y++) {
        for (var x = 0; x < imageData.width; x++) {
          if (x !== 0) patternFileString += ' '

          var offset = y * imageData.width * 4 + x * 4 + channelOffset
          var value = imageData.data[offset]

          patternFileString += String(value).padStart(3)
        }
        patternFileString += '\n'
      }
    }
  }
  return patternFileString
}

THREEx.ArPatternFile.triggerDownload = (patternFileString, newUrl) => {
  var domElement = window.document.createElement('a')
  domElement.href = window.URL.createObjectURL(
    new Blob([patternFileString], { type: 'text/plain' })
  )
  domElement.download = 'pattern-marker.patt'
  document.body.appendChild(domElement)
  domElement.click()
  document.body.removeChild(domElement)
  console.log('donlot')
  var e = document.createElement('a')
  var href = newUrl
  e.setAttribute('href', href)
  e.setAttribute('download', 'image.png')
  document.body.appendChild(e)
  e.click()
  document.body.removeChild(e)
}

THREEx.ArPatternFile.buildFullMarker = function (innerImageURL, onComplete) {
  var whiteMargin = 0.1
  var blackMargin = 0.2
  var innerMargin = whiteMargin + blackMargin

  var canvas = document.createElement('canvas')
  var context = canvas.getContext('2d')
  canvas.width = canvas.height = 512

  context.fillStyle = 'white'
  context.fillRect(0, 0, canvas.width, canvas.height)

  context.fillStyle = 'black'
  context.fillRect(
    whiteMargin * canvas.width,
    whiteMargin * canvas.height,
    canvas.width * (1 - 2 * whiteMargin),
    canvas.height * (1 - 2 * whiteMargin)
  )

  context.fillStyle = 'white'
  context.fillRect(
    innerMargin * canvas.width,
    innerMargin * canvas.height,
    canvas.width * (1 - 2 * innerMargin),
    canvas.height * (1 - 2 * innerMargin)
  )

  var innerImage = new Image()
  innerImage.src = innerImageURL
  innerImage.onload = async function (e) {
    await context.drawImage(
      innerImage,
      innerMargin * canvas.width,
      innerMargin * canvas.height,
      canvas.width * (1 - 2 * innerMargin),
      canvas.height * (1 - 2 * innerMargin)
    )
    var imageUrl = canvas.toDataURL()
    onComplete(imageUrl)
  }
  innerImage.onload()
}

class UserScreen extends Component {
  constructor () {
    super()
    this.state = {
      imageResult: '', // Image base64 without border for patt file
      patternFileStr: '', // Pattern string for patt file
      patternImage: '' // Image with border to download
    }
    this.handleUpload = this.handleUpload.bind(this)
    this.handleDownload = this.handleDownload.bind(this)
    this.encode = this.encode.bind(this)
  }

  handleUpload (e) {
    var reader = new FileReader()

    var self = this
    reader.onloadend = function () {
      THREEx.ArPatternFile.buildFullMarker(reader.result, function onComplete (
        patternResult
      ) {
        self.setState(
          {
            imageResult: reader.result,
            patternImage: patternResult
          },
          self.encode
        )
      })
    }
    reader.readAsDataURL(e.target.files[0])
  }
  encode () {
    let imageResult = this.state.imageResult
    var self = this
    THREEx.ArPatternFile.encodeImageUrl(imageResult, function onComplete (
      patternFileString
    ) {
      self.setState({
        patternFileStr: patternFileString
      })
    })
  }

  handleDownload () {
    let patternImage = this.state.patternImage
    let patternFileStr = this.state.patternFileStr
    THREEx.ArPatternFile.triggerDownload(patternFileStr, patternImage)
  }

  render () {
    return (
      <div>
        <Header location={this.props.location.pathname} />
        <div style={styles.container}>
          <TextField
            id='title'
            label='Title'
            helperText='ex: Your Title'
            fullWidth
            margin='normal'
          />
          <TextField
            id='detail'
            label='Detail'
            helperText='ex: Description Detail'
            fullWidth
            margin='normal'
          />
          <input
            accept='image/*'
            type='file'
            id='marker'
            name='marker'
            style={{ display: 'none' }}
            onChange={this.handleUpload}
          />
          <label htmlFor='marker'>
            <Button variant='raised' component='span' syle={styles.button}>
              Marker Upload
            </Button>
          </label>
          <input
            type='file'
            id='object'
            placeholder='object'
            style={{ display: 'none' }}
          />
          <label htmlFor='object'>
            <br />
            <Button variant='raised' component='span' syle={styles.button}>
              Object Upload
            </Button>
          </label>
          <div>
            <br />
            <Button
              variant='raised'
              component='span'
              color='primary'
              syle={styles.button}
            >
              Save
            </Button>
            {this.state.patternImage &&
              this.state.patternFileStr &&
              <Button
                variant='raised'
                component='span'
                color='secondary'
                onClick={this.handleDownload}
              >
                download
              </Button>}
          </div>
        </div>
        <Navigation />
      </div>
    )
  }
}

const styles = {
  container: {
    display: 'flex',
    flexDirection: 'column',
    flexWrap: 'wrap',
    marginTop: 56,
    padding: '2em'
  },
  button: {
    padding: '0.75em'
  }
}

export default UserScreen
