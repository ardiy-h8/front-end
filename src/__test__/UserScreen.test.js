import React from 'react'

import UserScreen from '../components/UserScreen'

import './__setupTest__.js'


var userScreen = new UserScreen()

describe('marker generator', () => {
  const getComponent = () => shallow(<UserScreen />)
  it('should be defined', () => {
    expect(getComponent()).toBeDefined()
  })

  it('should render button', () => {
    expect(getComponent()).toMatchSnapshot()
  })
  
  it('input should changes', () => {
    const wrapper = shallow(
      <input
        accept='image/*'
        type='file'
        id='marker'
        name='marker'
        style={{ display: 'none' }}
        onChange={userScreen.handleUpload}
      />
    )
    wrapper.find('input')
      .simulate('change', {target: {
        files: [new Blob(['../assets/hiro.png'], {type:'mime'})]
      }
    })
  })
})
