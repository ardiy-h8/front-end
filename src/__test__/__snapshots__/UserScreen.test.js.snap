// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`marker generator should render button 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <UserScreen />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Header />,
        <div
          style={
                    Object {
                              "display": "flex",
                              "flexDirection": "column",
                              "flexWrap": "wrap",
                              "marginTop": 56,
                              "padding": "2em",
                            }
          }
>
          <TextField
                    fullWidth={true}
                    helperText="ex: Your Title"
                    id="title"
                    label="Title"
                    margin="normal"
                    required={false}
                    select={false}
          />
          <TextField
                    fullWidth={true}
                    helperText="ex: Description Detail"
                    id="detail"
                    label="Detail"
                    margin="normal"
                    required={false}
                    select={false}
          />
          <input
                    accept="image/*"
                    id="marker"
                    name="marker"
                    onChange={[Function]}
                    style={
                              Object {
                                        "display": "none",
                                      }
                    }
                    type="file"
          />
          <label
                    htmlFor="marker"
          >
                    <WithStyles(Button)
                              component="span"
                              syle={
                                        Object {
                                                  "padding": "0.75em",
                                                }
                              }
                              variant="raised"
                    >
                              Marker Upload
                    </WithStyles(Button)>
          </label>
          <input
                    id="object"
                    placeholder="object"
                    style={
                              Object {
                                        "display": "none",
                                      }
                    }
                    type="file"
          />
          <label
                    htmlFor="object"
          >
                    <br />
                    <WithStyles(Button)
                              component="span"
                              syle={
                                        Object {
                                                  "padding": "0.75em",
                                                }
                              }
                              variant="raised"
                    >
                              Object Upload
                    </WithStyles(Button)>
          </label>
          <div>
                    <br />
                    <WithStyles(Button)
                              color="primary"
                              component="span"
                              syle={
                                        Object {
                                                  "padding": "0.75em",
                                                }
                              }
                              variant="raised"
                    >
                              Save
                    </WithStyles(Button)>
                    
          </div>
</div>,
        <Navigation />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <TextField
              fullWidth={true}
              helperText="ex: Your Title"
              id="title"
              label="Title"
              margin="normal"
              required={false}
              select={false}
/>,
            <TextField
              fullWidth={true}
              helperText="ex: Description Detail"
              id="detail"
              label="Detail"
              margin="normal"
              required={false}
              select={false}
/>,
            <input
              accept="image/*"
              id="marker"
              name="marker"
              onChange={[Function]}
              style={
                            Object {
                                          "display": "none",
                                        }
              }
              type="file"
/>,
            <label
              htmlFor="marker"
>
              <WithStyles(Button)
                            component="span"
                            syle={
                                          Object {
                                                        "padding": "0.75em",
                                                      }
                            }
                            variant="raised"
              >
                            Marker Upload
              </WithStyles(Button)>
</label>,
            <input
              id="object"
              placeholder="object"
              style={
                            Object {
                                          "display": "none",
                                        }
              }
              type="file"
/>,
            <label
              htmlFor="object"
>
              <br />
              <WithStyles(Button)
                            component="span"
                            syle={
                                          Object {
                                                        "padding": "0.75em",
                                                      }
                            }
                            variant="raised"
              >
                            Object Upload
              </WithStyles(Button)>
</label>,
            <div>
              <br />
              <WithStyles(Button)
                            color="primary"
                            component="span"
                            syle={
                                          Object {
                                                        "padding": "0.75em",
                                                      }
                            }
                            variant="raised"
              >
                            Save
              </WithStyles(Button)>
              
</div>,
          ],
          "style": Object {
            "display": "flex",
            "flexDirection": "column",
            "flexWrap": "wrap",
            "marginTop": 56,
            "padding": "2em",
          },
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "fullWidth": true,
              "helperText": "ex: Your Title",
              "id": "title",
              "label": "Title",
              "margin": "normal",
              "required": false,
              "select": false,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "fullWidth": true,
              "helperText": "ex: Description Detail",
              "id": "detail",
              "label": "Detail",
              "margin": "normal",
              "required": false,
              "select": false,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "accept": "image/*",
              "id": "marker",
              "name": "marker",
              "onChange": [Function],
              "style": Object {
                "display": "none",
              },
              "type": "file",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <WithStyles(Button)
                component="span"
                syle={
                                Object {
                                                "padding": "0.75em",
                                              }
                }
                variant="raised"
>
                Marker Upload
</WithStyles(Button)>,
              "htmlFor": "marker",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "Marker Upload",
                "component": "span",
                "syle": Object {
                  "padding": "0.75em",
                },
                "variant": "raised",
              },
              "ref": null,
              "rendered": "Marker Upload",
              "type": [Function],
            },
            "type": "label",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "id": "object",
              "placeholder": "object",
              "style": Object {
                "display": "none",
              },
              "type": "file",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <br />,
                <WithStyles(Button)
                  component="span"
                  syle={
                                    Object {
                                                      "padding": "0.75em",
                                                    }
                  }
                  variant="raised"
>
                  Object Upload
</WithStyles(Button)>,
              ],
              "htmlFor": "object",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {},
                "ref": null,
                "rendered": null,
                "type": "br",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": "Object Upload",
                  "component": "span",
                  "syle": Object {
                    "padding": "0.75em",
                  },
                  "variant": "raised",
                },
                "ref": null,
                "rendered": "Object Upload",
                "type": [Function],
              },
            ],
            "type": "label",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <br />,
                <WithStyles(Button)
                  color="primary"
                  component="span"
                  syle={
                                    Object {
                                                      "padding": "0.75em",
                                                    }
                  }
                  variant="raised"
>
                  Save
</WithStyles(Button)>,
                "",
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {},
                "ref": null,
                "rendered": null,
                "type": "br",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": "Save",
                  "color": "primary",
                  "component": "span",
                  "syle": Object {
                    "padding": "0.75em",
                  },
                  "variant": "raised",
                },
                "ref": null,
                "rendered": "Save",
                "type": [Function],
              },
              "",
            ],
            "type": "div",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Header />,
          <div
            style={
                        Object {
                                    "display": "flex",
                                    "flexDirection": "column",
                                    "flexWrap": "wrap",
                                    "marginTop": 56,
                                    "padding": "2em",
                                  }
            }
>
            <TextField
                        fullWidth={true}
                        helperText="ex: Your Title"
                        id="title"
                        label="Title"
                        margin="normal"
                        required={false}
                        select={false}
            />
            <TextField
                        fullWidth={true}
                        helperText="ex: Description Detail"
                        id="detail"
                        label="Detail"
                        margin="normal"
                        required={false}
                        select={false}
            />
            <input
                        accept="image/*"
                        id="marker"
                        name="marker"
                        onChange={[Function]}
                        style={
                                    Object {
                                                "display": "none",
                                              }
                        }
                        type="file"
            />
            <label
                        htmlFor="marker"
            >
                        <WithStyles(Button)
                                    component="span"
                                    syle={
                                                Object {
                                                            "padding": "0.75em",
                                                          }
                                    }
                                    variant="raised"
                        >
                                    Marker Upload
                        </WithStyles(Button)>
            </label>
            <input
                        id="object"
                        placeholder="object"
                        style={
                                    Object {
                                                "display": "none",
                                              }
                        }
                        type="file"
            />
            <label
                        htmlFor="object"
            >
                        <br />
                        <WithStyles(Button)
                                    component="span"
                                    syle={
                                                Object {
                                                            "padding": "0.75em",
                                                          }
                                    }
                                    variant="raised"
                        >
                                    Object Upload
                        </WithStyles(Button)>
            </label>
            <div>
                        <br />
                        <WithStyles(Button)
                                    color="primary"
                                    component="span"
                                    syle={
                                                Object {
                                                            "padding": "0.75em",
                                                          }
                                    }
                                    variant="raised"
                        >
                                    Save
                        </WithStyles(Button)>
                        
            </div>
</div>,
          <Navigation />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <TextField
                fullWidth={true}
                helperText="ex: Your Title"
                id="title"
                label="Title"
                margin="normal"
                required={false}
                select={false}
/>,
              <TextField
                fullWidth={true}
                helperText="ex: Description Detail"
                id="detail"
                label="Detail"
                margin="normal"
                required={false}
                select={false}
/>,
              <input
                accept="image/*"
                id="marker"
                name="marker"
                onChange={[Function]}
                style={
                                Object {
                                                "display": "none",
                                              }
                }
                type="file"
/>,
              <label
                htmlFor="marker"
>
                <WithStyles(Button)
                                component="span"
                                syle={
                                                Object {
                                                                "padding": "0.75em",
                                                              }
                                }
                                variant="raised"
                >
                                Marker Upload
                </WithStyles(Button)>
</label>,
              <input
                id="object"
                placeholder="object"
                style={
                                Object {
                                                "display": "none",
                                              }
                }
                type="file"
/>,
              <label
                htmlFor="object"
>
                <br />
                <WithStyles(Button)
                                component="span"
                                syle={
                                                Object {
                                                                "padding": "0.75em",
                                                              }
                                }
                                variant="raised"
                >
                                Object Upload
                </WithStyles(Button)>
</label>,
              <div>
                <br />
                <WithStyles(Button)
                                color="primary"
                                component="span"
                                syle={
                                                Object {
                                                                "padding": "0.75em",
                                                              }
                                }
                                variant="raised"
                >
                                Save
                </WithStyles(Button)>
                
</div>,
            ],
            "style": Object {
              "display": "flex",
              "flexDirection": "column",
              "flexWrap": "wrap",
              "marginTop": 56,
              "padding": "2em",
            },
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "fullWidth": true,
                "helperText": "ex: Your Title",
                "id": "title",
                "label": "Title",
                "margin": "normal",
                "required": false,
                "select": false,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "fullWidth": true,
                "helperText": "ex: Description Detail",
                "id": "detail",
                "label": "Detail",
                "margin": "normal",
                "required": false,
                "select": false,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "accept": "image/*",
                "id": "marker",
                "name": "marker",
                "onChange": [Function],
                "style": Object {
                  "display": "none",
                },
                "type": "file",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <WithStyles(Button)
                  component="span"
                  syle={
                                    Object {
                                                      "padding": "0.75em",
                                                    }
                  }
                  variant="raised"
>
                  Marker Upload
</WithStyles(Button)>,
                "htmlFor": "marker",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": "Marker Upload",
                  "component": "span",
                  "syle": Object {
                    "padding": "0.75em",
                  },
                  "variant": "raised",
                },
                "ref": null,
                "rendered": "Marker Upload",
                "type": [Function],
              },
              "type": "label",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "id": "object",
                "placeholder": "object",
                "style": Object {
                  "display": "none",
                },
                "type": "file",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <br />,
                  <WithStyles(Button)
                    component="span"
                    syle={
                                        Object {
                                                            "padding": "0.75em",
                                                          }
                    }
                    variant="raised"
>
                    Object Upload
</WithStyles(Button)>,
                ],
                "htmlFor": "object",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {},
                  "ref": null,
                  "rendered": null,
                  "type": "br",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "children": "Object Upload",
                    "component": "span",
                    "syle": Object {
                      "padding": "0.75em",
                    },
                    "variant": "raised",
                  },
                  "ref": null,
                  "rendered": "Object Upload",
                  "type": [Function],
                },
              ],
              "type": "label",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <br />,
                  <WithStyles(Button)
                    color="primary"
                    component="span"
                    syle={
                                        Object {
                                                            "padding": "0.75em",
                                                          }
                    }
                    variant="raised"
>
                    Save
</WithStyles(Button)>,
                  "",
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {},
                  "ref": null,
                  "rendered": null,
                  "type": "br",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "children": "Save",
                    "color": "primary",
                    "component": "span",
                    "syle": Object {
                      "padding": "0.75em",
                    },
                    "variant": "raised",
                  },
                  "ref": null,
                  "rendered": "Save",
                  "type": [Function],
                },
                "",
              ],
              "type": "div",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
